### PR TITLE
Fixed learner search for DEDP fails issues

### PR DESCRIPTION
**Pull request by amir-qayyum-khan**
#### What are the relevant tickets?
- Fixes https://github.com/mitodl/micromasters/issues/2273

#### how to reproduce?
- To reproduce it just make country code of a profile null. This error is because of data inconsistency 

@pdpinch

### CODE REVIEW

**Comment by amir-qayyum-khan** [comment]
Fixed

```js
console.js:26 TypeError: Cannot read property 'trim' of null
    at Object.country (iso3166.min.js:9)
    at d (util.js:282)
    at t.value (LearnerResult.js:51)
    at p._renderValidatedComponentWithoutOwnerOrContext (ReactCompositeComponent.js:799)
    at p._renderValidatedComponent (ReactCompositeComponent.js:822)
    at performInitialMount (ReactCompositeComponent.js:362)
    at p.mountComponent (ReactCompositeComponent.js:258)
    at Object.mountComponent (ReactReconciler.js:46)
    at performInitialMount (ReactCompositeComponent.js:371)
    at p.mountComponent (ReactCompositeComponent.js:258)
    at Object.mountComponent (ReactReconciler.js:46)
    at mountChildren (ReactMultiChild.js:238)
    at m._createInitialChildren (ReactDOMComponent.js:691)
    at mountComponent (ReactDOMComponent.js:516)
    at Object.mountComponent (ReactReconciler.js:46)
    at performInitialMount (ReactCompositeComponent.js:371)
    at p.mountComponent (ReactCompositeComponent.js:258)
    at Object.mountComponent (ReactReconciler.js:46)
    at p._updateRenderedComponent (ReactCompositeComponent.js:765)
    at p._performComponentUpdate (ReactCompositeComponent.js:724)
    at p.updateComponent (ReactCompositeComponent.js:645)
    at p.receiveComponent (ReactCompositeComponent.js:547)
    at Object.receiveComponent (ReactReconciler.js:125)
    at Object.updateChildren (ReactChildReconciler.js:109)
    at m._reconcilerUpdateChildren (ReactMultiChild.js:213)
    at m._updateChildren (ReactMultiChild.js:312)
    at m.updateChildren (ReactMultiChild.js:299)
    at m._updateDOMChildren (ReactDOMComponent.js:930)
    at m.updateComponent (ReactDOMComponent.js:748)
    at m.receiveComponent (ReactDOMComponent.js:710)
    at Object.receiveComponent (ReactReconciler.js:125)
    at p._updateRenderedComponent (ReactCompositeComponent.js:754)
    at p._performComponentUpdate (ReactCompositeComponent.js:724)
    at p.updateComponent (ReactCompositeComponent.js:645)
    at p.receiveComponent (ReactCompositeComponent.js:547)
    at Object.receiveComponent (ReactReconciler.js:125)
    at Object.updateChildren (ReactChildReconciler.js:109)
    at m._reconcilerUpdateChildren (ReactMultiChild.js:213)
    at m._updateChildren (ReactMultiChild.js:312)
    at m.updateChildren (ReactMultiChild.js:299)
    at m._updateDOMChildren (ReactDOMComponent.js:930)
    at m.updateComponent (ReactDOMComponent.js:748)
    at m.receiveComponent (ReactDOMComponent.js:710)
    at Object.receiveComponent (ReactReconciler.js:125)
    at p._updateRenderedComponent (ReactCompositeComponent.js:754)
    at p._performComponentUpdate (ReactCompositeComponent.js:724)
    at p.updateComponent (ReactCompositeComponent.js:645)
    at p.receiveComponent (ReactCompositeComponent.js:547)
    at Object.receiveComponent (ReactReconciler.js:125)
    at Object.updateChildren (ReactChildReconciler.js:109)
t.(anonymous function) @ console.js:26
e.setError @ SearchkitManager.js:207
e.setError @ SearchRequest.js:22
```

**Comment by aliceriot** [comment]
Amir, can you give a minimal example of how to reproduce the error (that will work on `master` right now) ?

**Comment by amir-qayyum-khan** [comment]
updated description of pr

**Comment by aliceriot** [comment]
thanks, will take a look in a sec

**Comment by pdpinch** [comment]
Amir, do you know what data in particular is bad in production? This isn't a blocking question, but we may have to do some investigation to understand the root cause of the bad data in addition to this front-end change.

**Comment by amir-qayyum-khan** [comment]
@pdpinch it is profile/country code which is creating issue

**Comment by rhysyngsun** [comment]
@amir-qayyum-khan we ended up removing the user's enrollment and re-indexing for an immediate fix.

File: static/js/components/search/CountryRefinementOption.js [hunk 1]
```diff
@@ -10,17 +10,22 @@ export default class CountryRefinementOption extends React.Component {
...
-  countryName = (countryCode: string): string => (
-    iso3166.country(countryCode).name
-  );
+  countryName = (countryCode: string): string => {
+    let country, countryName = "-";
+    if (countryCode) {
+      country = iso3166.country(countryCode);
+      countryName = country ? country.name : "-";
+    }
+    return countryName;
```
Comment (aliceriot) [inline static/js/components/search/CountryRefinementOption.js]: Why do we want these to default to `"-"` here? just wondering.

It might be a little more elegant to wrap the call to `iso3166.country` in `R.defaultTo`, which will return it's first argument if the second one if `null`, `undefined`, or `NaN`. So you could just do:

```js
let country = R.defaultTo("-", iso3166.country(countryCode));
return country === "-" ? country : country.name;
```

something like that!

File: static/js/util/util.js [hunk 2]
```diff
@@ -272,14 +272,17 @@ export function getPreferredName(profile: Profile, last: boolean = true): string
  */
 export function getLocation(profile: Profile, showState: boolean = true): string {
   let { country, state_or_territory, city } = profile;
-  let subCountryLocation, countryLocation;
+  let subCountryLocation = city || "";
```
Comment (aliceriot) [inline static/js/util/util.js]: why did you move this `subCountryLocation` statement out of the `else` block below?

File: static/js/util/util.js [hunk 3]
```diff
@@ -272,14 +272,17 @@ export function getPreferredName(profile: Profile, last: boolean = true): string
...
     let state = state_or_territory.replace(/^\D{2}-/, '');
     subCountryLocation = showState ? `${city}, ${state}` : city;
     countryLocation = 'US';
-  } else {
-    subCountryLocation = city;
-    countryLocation = iso3166.country(country).name;
+  } else if (country) {
+    countryFromCode = iso3166.country(country);
+    countryLocation = countryFromCode ? countryFromCode.name : "-";
   }
```
Comment (aliceriot) [inline static/js/util/util.js]: can you add back an `else` block that will perform the default case?

File: static/js/components/search/CountryRefinementOption.js (hunk 1 above)
Comment (amir-qayyum-khan) [inline static/js/components/search/CountryRefinementOption.js]: the issue is this iso3166.country(countryCode), it will crash if ``countryCode`` is not valid.

```js
TypeError: Cannot read property 'trim' of null
      at Object.functions.country (node_modules/iso-3166-2/iso3166.min.js:9:17511)
      at CountryRefinementOption.__countryName__REACT_HOT_LOADER__ (static/js/components/search/CountryRefinementOption.js:15:44)
      at CountryRefinementOption._this.countryName (static/js/components/search/CountryRefinementOption.js:46:65)
      at CountryRefinementOption.render (static/js/components/search/CountryRefinementOption.js:34:18)
      at node_modules/react-dom/lib/ReactCompositeComponent.js:796:21
      at measureLifeCyclePerf (node_modules/react-dom/lib/ReactCompositeComponent.js:75:12)
      at ReactCompositeComponent._renderValidatedComponentWithoutOwnerOrContext (node_modules/react-dom/lib/ReactCompositeComponent.js:795:25)
      at ReactCompositeComponent._renderValidatedComponent (node_modules/react-dom/lib/ReactCompositeComponent.js:822:32)
      at ReactCompositeComponent.performInitialMount (node_modules/react-dom/lib/ReactCompositeComponent.js:362:30)
      at ReactCompositeComponent.mountComponent (node_modules/react-dom/lib/ReactCompositeComponent.js:258:21)
      at Object.ReactReconciler.mountComponent (node_modules/react-dom/lib/ReactReconciler.js:46:35)
      at ReactCompositeComponent.performInitialMount (node_modules/react-dom/lib/ReactCompositeComponent.js:371:34)
      at ReactCompositeComponent.mountComponent (node_modules/react-dom/lib/ReactCompositeComponent.js:258:21)
      at Object.ReactReconciler.mountComponent (node_modules/react-dom/lib/ReactReconciler.js:46:35)
      at mountComponentIntoNode (node_modules/react-dom/lib/ReactMount.js:104:32)
      at ReactReconcileTransaction.TransactionImpl.perform (node_modules/react-dom/lib/Transaction.js:140:20)
      at batchedMountComponentIntoNode (node_modules/react-dom/lib/ReactMount.js:126:15)
      at ReactDefaultBatchingStrategyTransaction.TransactionImpl.perform (node_modules/react-dom/lib/Transaction.js:140:20)
      at Object.ReactDefaultBatchingStrategy.batchedUpdates (node_modules/react-dom/lib/ReactDefaultBatchingStrategy.js:62:26)
      at Object.batchedUpdates (node_modules/react-dom/lib/ReactUpdates.js:97:27)
      at Object.ReactMount._renderNewRootComponent (node_modules/react-dom/lib/ReactMount.js:320:18)
      at Object.ReactMount._renderSubtreeIntoContainer (node_modules/react-dom/lib/ReactMount.js:401:32)
      at Object.ReactMount.render (node_modules/react-dom/lib/ReactMount.js:422:23)
      at makeStrippedHtml (static/js/util/util.js:243:14)
      at renderCountryOption (static/js/components/search/CountryRefinementOption_test.js:14:5)
      at Context.<anonymous> (static/js/components/search/CountryRefinementOption_test.js:37:18)


```

**Comment by amir-qayyum-khan** [comment]
Done @aliceriot

File: static/js/components/search/CountryRefinementOption.js (hunk 1 above)
Comment (aliceriot) [inline static/js/components/search/CountryRefinementOption.js]: Ah, it will do that if the country code is `null` or `undefined`. if you call `iso3166.country('')` it just returns an empty object though, so maybe instead we could do:

```js
let country = iso3166.country(R.defaultTo("", countryCode)); 
return R.pathOr("-", ['name'], country);
```

alternatively, there's a little helper function in `lib/currency.js` that should be rewritten! It would also have this bug. We could rewrite it as:

```js
const codeToCountryName = R.compose(
  R.pathOr("-", ['name']), iso3166.country, R.defaultTo("")
);
```

You might need to make slight changes to that, but I think it should work. Then you could just import that function here and use it (and maybe also in the other file you changed for this PR).

Comment (aliceriot) [inline static/js/components/search/CountryRefinementOption.js]: Oh also, why `"-"` and not just `""`?

Comment (aliceriot) [inline static/js/components/search/CountryRefinementOption.js]: ^^ I think the best thing would be to fix the function in `lib/currency` and then use it here (also, add a test for the `codeToCountryName` function which tests the case when the code is `null` or `undefined`).

**Comment by amir-qayyum-khan** [comment]
Done @aliceriot

File: static/js/lib/currency.js [hunk 4]
```diff
@@ -37,7 +37,9 @@ const codesToOptions = R.compose(
 
 export const currencyOptions = codesToOptions(cc.codes());
 
-export const codeToCountryName = (code: string) => iso3166.country(code).name || '';
+export const codeToCountryName = (code: string) =>  R.pathOr(
+  "N/A", ['name'], iso3166.country(R.defaultTo("", code))
+);
```
Comment (aliceriot) [inline static/js/lib/currency.js]: could we make this function return `''` instead of `'N/A'` and only display `'N/A'` in the `countryRefinementOption`? I think that the change to `'N/A'` might mess up the `currencyForCountry` function below.

Comment (aliceriot) [inline static/js/lib/currency.js]: In `countryRefinementOption` you could do something like

```js
R.when(R.equals(""), () => 'N/A', codeToCountryName(country))
```

**Comment by amir-qayyum-khan** [comment]
Doen @aliceriot

File: static/js/util/util.js [hunk 5]
```diff
@@ -273,15 +273,21 @@ export function getPreferredName(profile: Profile, last: boolean = true): string
...
+  city = city ? `${city}, ` : "";
+
   if ( country === 'US' ) {
     let state = state_or_territory.replace(/^\D{2}-/, '');
-    subCountryLocation = showState ? `${city}, ${state}` : city;
+    subCountryLocation = showState ? `${city}${state}, ` : city;
     countryLocation = 'US';
+  } else if (country) {
+    subCountryLocation = city;
+    let countryFromCode = iso3166.country(country);
```
Comment (aliceriot) [inline static/js/util/util.js]: I think we should use `codeToCountryName` here as wlel

File: static/js/util/util_test.js [hunk 6]
```diff
@@ -179,6 +179,33 @@ describe('utility functions', () => {
       // assert hide state
       assert.equal(getLocation(us, false), 'Portland, US');
     });
+
+    it('should return `${city}, ` when no counrty code', () => {
```
Comment (noisecapella) [inline static/js/util/util_test.js]: `counrty` -> `country`

File: static/js/util/util_test.js [hunk 7]
```diff
@@ -179,6 +179,33 @@ describe('utility functions', () => {
...
+    it('should return `US` when no city', () => {
+      let us = {
+        country: 'US',
+        state_or_territory: 'US-ME',
+        city: null
+      };
+      assert.equal(getLocation(us, false), 'US');
+    });
+
+    it('should return `-` when no city and no counrty', () => {
```
Comment (noisecapella) [inline static/js/util/util_test.js]: `counrty` -> `country`

File: static/js/util/util.js (hunk 5 above)
Comment (aliceriot) [inline static/js/util/util.js]: In fact, given that `codeToCountryName` is now null-safe you should just be able to do collapse the `if else` and `else` blocks like this:

```js
else {
  subCountryLocation = city;
  countryLocation = codeToCountryName(country);
}
```

File: static/js/util/util_test.js (hunk 6 above)
Comment (aliceriot) [inline static/js/util/util_test.js]: mis-spelling

File: static/js/util/util_test.js (hunk 7 above)
Comment (aliceriot) [inline static/js/util/util_test.js]: another mis-spelling

**Comment by amir-qayyum-khan** [comment]
Done @aliceriot

File: static/js/lib/currency.js [hunk 8]
```diff
@@ -25,6 +23,8 @@ export const excludedCurrencyCodes = [
   'XXX'
 ];
 
+const labelSort = R.sortBy(R.compose(R.toLower, R.prop('label')));
```
Comment (aliceriot) [inline static/js/lib/currency.js]: Instead of this, can we make a new file (`lib/location.js`) and put both `getLocation` (from `util.js`) and `codeToCountryName` in there?

File: static/js/lib/currency.js [hunk 9]
```diff
@@ -37,9 +37,7 @@ const codesToOptions = R.compose(
 
 export const currencyOptions = codesToOptions(cc.codes());
 
-export const codeToCountryName = (code: string) =>  R.pathOr(
-  "", ['name'], iso3166.country(R.defaultTo("", code))
-);
+
```
Comment (aliceriot) [inline static/js/lib/currency.js]: extra blank line

**Comment by amir-qayyum-khan** [comment]
Done @aliceriot
